File: cpp/endpoints/localization/convert_ll_to_xy.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_localization_msgs/srv/convert_lat_lon_to_cartesian.hpp"
#include "sensor_msgs/msg/nav_sat_fix.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

// TODO: Enter your latitude/longitude
const double LATITUDE = 0.0;
const double LONGITUDE = 0.0;

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("convert_lat_lon_to_cartesian_client");
  rclcpp::Client<clearpath_localization_msgs::srv::ConvertLatLonToCartesian>::SharedPtr client =
    node->create_client<clearpath_localization_msgs::srv::ConvertLatLonToCartesian>("/" + namespace_ + "/" + std::string("localization/lat_lon_to_xy"));

  auto request = std::make_shared<clearpath_localization_msgs::srv::ConvertLatLonToCartesian::Request>();
  auto lat_lon_msg = sensor_msgs::msg::NavSatFix();
  lat_lon_msg.latitude = LATITUDE;
  lat_lon_msg.longitude = LONGITUDE;
  request->msg = lat_lon_msg;

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("convert_lat_lon_to_cartesian_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("convert_lat_lon_to_cartesian_client"), "/%s/localization/lat_lon_to_xy service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("convert_lat_lon_to_cartesian_client"), "Converted lat/lon to Cartesian successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("convert_lat_lon_to_cartesian_client"), "Failed to convert lat/lon to Cartesian");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("convert_lat_lon_to_cartesian_client"), "Failed to call service /%s/localization/lat_lon_to_xy", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}

File: python/endpoints/localization/convert_ll_to_xy.py
from clearpath_config.common.utils.yaml import read_yaml
from clearpath_localization_msgs.srv import ConvertLatLonToCartesian
from sensor_msgs.msg import NavSatFix
import rclpy
from rclpy.node import Node

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ConvertLatLonToCartesian'

# TODO: Enter your lat/lon coordinates
LAT_LON_MSG = NavSatFix(
            latitude=,
            longitude=,
)

class ConvertLatLonToCartesianNode(Node):

    def __init__(self, namespace: str):
        super().__init__('convert_lat_lon_to_cartesian')
        self._convert_client = self.create_client(ConvertLatLonToCartesian, f'/{namespace}/localization/lat_lon_to_xy')

        while not self._convert_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/localization/lat_lon_to_xy service not available, waiting again...')

    def convert(self, msg):
        req = ConvertLatLonToCartesian.Request()
        req.msg = msg
        self.future = self._convert_client.call_async(req)
        self.future.add_done_callback(self.convert_srv_response_cb)

    def convert_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Converted lat/lon to Cartesian successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to convert lat/lon to Cartesian')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    convert_client = ConvertLatLonToCartesianNode(namespace)
    convert_client.convert(LAT_LON_MSG)

    rclpy.spin(convert_client)


if __name__ == '__main__':
    main()
